pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -326,6 +328,25 @@ type attachDetachController struct {
 
 	// pvcQueue is used to queue pvc objects
 	pvcQueue workqueue.RateLimitingInterface
+
+	// operationTimestamps caches start timestamp of attach/detach for metric recording.
+	// Detailed lifecyle/key for each operation
+	// 1. attach
+	//     key:        string of VolumeToAttach.NodeName + "/" + VolumeToAttach.VolumeName
+	//     start time: the first time "AttachVolume" is called in reconciler in "attachDesiredVolumes"
+	//                 for the VolumeToAttach object
+	//     end time:   after the "VolumeToAttach" has been successfully attached to the desired node
+	//     abort:      The "VolumeToAttach" has been marked "attached", i.e., actualStateOfWorld
+	//                 returns true on "IsVolumeAttachedToNode", before a successful "AttachVolume" call
+	//                 has been invoked by the reconciler
+	// 2. detach
+	//     key:        string of AttachedVolume.NodeName + "/" + AttachedVolume.VolumeName
+	//     start time: the first time "DetachVolume" is called in reconciler in "reconcile"
+	//                 for the "AttachedVolume" object
+	//     end time:   after the volume "AttachedVolume" object refers to being successfully
+	//                 detached from the node
+	//     abort:      N.A.
```

Comment:
@gnufied @jingxu97 is it possible to "abort" detach if the volume gets added to desired state again?

Reply:
I think it is possible. the detach process first get the actual state, and cross compare to desired state, it will only do the detach operation if the volume does not appear in desired state. I will add logic to abort the operation under that circumstance to avoid possible memory leak. Any other scenarios? @gnufied @jingxu97